# LINEAR REGRESSION -  MINIPROYECTO

## FORMA 1 - GRADIENT DESCENT

El **descenso de gradiente** es un algoritmo de optimización que se utiliza para ajustar los parámetros de un modelo de regresión lineal. El objetivo es encontrar los valores de los parámetros que minimizan la función de costo, que mide la diferencia entre las predicciones del modelo y los valores reales.

El descenso de gradiente funciona iterativamente, actualizando los valores de los parámetros en cada iteración para reducir la función de costo. En cada iteración, se calcula el gradiente de la función de costo con respecto a los parámetros y se actualizan los parámetros en la dirección opuesta al gradiente. La tasa de aprendizaje controla el tamaño de los pasos que se dan en cada iteración.

Para aplicar el descenso de gradiente a un modelo de regresión lineal, se utiliza el **MSE** como función de costo. El MSE mide la diferencia entre las predicciones del modelo y los valores reales al cuadrado. El objetivo es minimizar el MSE ajustando los valores de los parámetros.


Paso a paso del algoritmo de descenso de gradiente para optimizar un modelo de regresión lineal:

1. Inicializa los parámetros del modelo con valores aleatorios.
2. Calcula la función de costo utilizando los valores actuales de los parámetros.
3. Calcula el gradiente de la función de costo con respecto a los parámetros.
4. Actualiza los parámetros en la dirección opuesta al gradiente multiplicado por la tasa de aprendizaje.
5. Repite los pasos 2-4 hasta que la función de costo converja a un mínimo o decidas parar.

COMO LO TENEIS QUE HACER VOSOTROS:

1. Inicializa los parametros `m` y `c` a cero o con valores aleatorios.
2. Elegimos un numero de iteraciones (eg. 100) y un ``learning_rate`` (eg. 0.01)
3. Inicializamos valores para X e Y. X que sean numpy arrays de 10 o 20 numeros (eg. range(10)) y la Y pueden ser parecida a la X pero con numeros arriba o abajo para simular aletoriedad (eg. 2,2,4,2,5,7,6,9,8,10)
4. Arrancamos el bucle for.
5. Calculamos la derivada del error respecto de `m` y `c`:
    - `dm = 2/n*sum((y-(m*x+c))*(-x))`
    - `dc = 2/n*sum((y-(m*x+c))*(-1))`
6. Actualizamos los parametros: `m=m-dm*lr` y `c=c-dc*lr`
7. Cada iteracion haceis un plot de la linea `x` y `m*x+c` respecto a los puntos originales (plot de x e y) y os guardais el las imagenes (Si haceis 100 iteraciones, tendreis 100 imagenes. Podeis valorar solo guardar por ejemplo cada 5 iteraciones por no hacer tantos plots).
8. Una vez terminamos printeamos m y c y con los plots generados teneis quer buscar alguna libreria en python y montar un gif con las imagenes de los plots:

![gif](https://th.bing.com/th/id/R.79e22f97090c346d704a68f7151e8cda?rik=oJV36GZyA1otdA&riu=http%3a%2f%2fcdn-images-1.medium.com%2fmax%2f640%2f1*eeIvlwkMNG1wSmj3FR6M2g.gif&ehk=0NUalJOl26VxY8ndNrkpV7GwYM1NVtJ5kMxU6jm5jB0%3d&risl=&pid=ImgRaw&r=0)

NOTA**: Todos los valores son orientativos. Podeis modificar los datos como querais. De hecho, os animo a que probeis y trasteeis.

In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [5]:
from IPython.display import clear_output

x=np.array([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20])
y=np.array([10,10,30,40,20,40,60,40,60,80,70,60,80,100,90,100,120,130,140,120]) #Definimos los datos: x e y son matrices NumPy que representan los puntos en el gráfico

m=0
c=0 #Inicializamos los parámetros: m y c representan la pendiente y la intersección en el eje y de la línea que estamos tratando de encontrar
lrate=0.001 #Establecemos la tasa de aprendizaje: lrate, controla cuánto queremos ajustar los parámetros en cada iteración del algoritmo
n = len(y

for i in range(100000): #Bucle de optimización: Aquí es donde ocurre la magia.
                        #Iteramos 100,000 veces para ajustar los valores de m y c para minimizar el error entre la línea que estamos ajustando y los puntos reales en el gráfico
   clear_output(wait=True)
   dm = 2/n*sum((y-(m*x+c))*(-x)) #Calculamos las derivadas
   dc = 2/n*sum((y-(m*x+c))*(-1))
   m = m - dm * lrate
   c = c - dc * lrate #Actualización de los parámetros: En cada iteración, calculamos el "gradiente" (la dirección en la que debemos ajustar m y c) y actualizamos los valores de m y c usando la tasa de aprendizaje
   
   fig = plt.Figure()
   plt.scatter(x,y)
   plt.plot(x, m*x+c, "r")
   plt.title(f"iter:{i}, error{(sum((y-(m*x+c))**2))/len(y)}")
   if i >= 0 and i <= 9:
      plt.savefig(f"iteration_00{i}.png")
      plt.show()
      plt.close()
   else:
      plt.savefig(f"iteration_0{i}.png")
      plt.show()
      plt.close()

#Visualización: Código para visualizar el progreso en cada iteración. Crea gráficos que muestran los puntos de datos reales y la línea que estamos ajustando. Vemos cómo la línea va mejorando con cada iteración

#Estamos buscando la mejor línea para ajustarse a los puntos de datos dados utilizando el método del descenso de gradiente, un algoritmo de optimización ampliamente utilizado en el aprendizaje automático y la optimización numérica

In [6]:
print(m)
print(c)

6.496240601503783
1.7894736842102228


In [17]:
import os
import imageio

In [40]:
path = 'fotos/' #Ruta donde se encuentran las imágenes
archivos = sorted(os.listdir(path)) #Lista que contiene los nombres de los archivos en el path (sorted)
img_array = []
 
for x in range(0, len(archivos)): #Bucle for para iterar sobre cada archivo en archivos
    nomArchivo = archivos[x] #Se construye ruta completa de cada archivo
    dirArchivo = path + str(nomArchivo)
    print(dirArchivo)
 
    leer_imagen = imageio.imread(dirArchivo) #Se utiliza la función imageio.imread() para leer cada imagen del archivo. La imagen se guarda en la variable leer_imagen
 
    img_array.append(leer_imagen) #Cada imagen leída se agrega a la lista img_array
 
imageio.mimwrite('Gradient_Descent.GIF', img_array, 'GIF', duration=0.5) #Escribir imagenes en un archivo GIF. (Se espicifica el nombre, se proporciona la lista de imgns, formato GIF, duracion de cada imagen)


fotos/iteration_000.png
fotos/iteration_001.png
fotos/iteration_002.png
fotos/iteration_003.png
fotos/iteration_004.png
fotos/iteration_005.png
fotos/iteration_006.png
fotos/iteration_007.png


C:\Users\igbopa\AppData\Local\Temp\ipykernel_25748\319375057.py:11: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  leer_imagen = imageio.imread(dirArchivo)


fotos/iteration_008.png
fotos/iteration_009.png
fotos/iteration_010.png
fotos/iteration_011.png
fotos/iteration_012.png
fotos/iteration_013.png
fotos/iteration_014.png
fotos/iteration_015.png
fotos/iteration_016.png
fotos/iteration_017.png
fotos/iteration_018.png
fotos/iteration_019.png
fotos/iteration_020.png
fotos/iteration_021.png
fotos/iteration_022.png
fotos/iteration_023.png
fotos/iteration_024.png
fotos/iteration_025.png
fotos/iteration_026.png
fotos/iteration_027.png
fotos/iteration_028.png
fotos/iteration_029.png
fotos/iteration_030.png
fotos/iteration_031.png
fotos/iteration_032.png
fotos/iteration_033.png
fotos/iteration_034.png
fotos/iteration_035.png
fotos/iteration_036.png
fotos/iteration_037.png
fotos/iteration_038.png
fotos/iteration_039.png
fotos/iteration_040.png
fotos/iteration_041.png
fotos/iteration_042.png
fotos/iteration_043.png
fotos/iteration_044.png
fotos/iteration_045.png
fotos/iteration_046.png
fotos/iteration_047.png
fotos/iteration_048.png
fotos/iteration_

# FORMA 2 - ECUACION NORMAL

El modelo de Linear Regression es unico en todo el universo ML. Es el unico que se puede resolver de forma analitica resolviendo una ecuacion sin necesidad de optimizar usando por ejemplo el descenso de gradiente. Para resolverlo se usa una ecuacion que se llama ECUACION NORMAL.

La ecuación normal es una forma de resolver la regresión lineal sin utilizar el descenso de gradiente. En lugar de iterar para encontrar los valores de los parámetros que minimizan la función de costo, la ecuación normal calcula los valores de los parámetros directamente.
La ecuación normal para la regresión lineal simple es:

$$\theta=(X^{T}X)^{−1}X^{T}y$$
Donde θ es el vector de parámetros, X es la matriz de características, y es el vector de valores objetivo y −1 denota la inversa de una matriz.
La ecuación normal se deriva al igualar el gradiente de la función de costo a cero. La solución resultante es la que minimiza la función de costo.

AQUI TENEIS LA FORMULA RESUELTA: https://interactivechaos.com/es/manual/tutorial-de-machine-learning/la-ecuacion-normal

Buscadla y aplicarla con vuestros datos de `X` e ``y``, y os calcula directamente los parametros ``m`` y ``c``.

In [24]:
X = pd.DataFrame(np.array([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]))
y = np.array([10,10,30,40,20,40,60,40,60,80,70,60,80,100,90,100,120,130,140,120]) #Definimos los datos: x e y son matrices NumPy que representan los puntos en el gráfico

#Fórmula analítica que calcula los valores óptimos de los coeficientes sin necesidad de iteraciones como con el anterior metodo

In [25]:
X[:5]

#En este momento la estructura X tiene esta forma:

0
0  1
1  2
2  3
3  4
4  5

In [27]:
X_1 = np.c_[np.ones((len(X), 1)), X]
X_1

#Hay que añadir el valor 1 (término independiente) a cada una de las muestras X generadas

#Creamos la nueva matriz, agregando la columna de unos al principio de la matriz x (Esto se hace para poder realizar cálculos de regresión lineal con un término de sesgo (intercepto))

#np.c_[] concatena estas matrices de unos con la matriz X

array([[ 1.,  1.],
       [ 1.,  2.],
       [ 1.,  3.],
       [ 1.,  4.],
       [ 1.,  5.],
       [ 1.,  6.],
       [ 1.,  7.],
       [ 1.,  8.],
       [ 1.,  9.],
       [ 1., 10.],
       [ 1., 11.],
       [ 1., 12.],
       [ 1., 13.],
       [ 1., 14.],
       [ 1., 15.],
       [ 1., 16.],
       [ 1., 17.],
       [ 1., 18.],
       [ 1., 19.],
       [ 1., 20.]])

In [29]:
best_W = np.linalg.inv(X_1.T.dot(X_1)).dot(X_1.T).dot(y)
best_W #Mejores coeficientes

#Este código realiza manipulaciones de matrices y cálculos matriciales para calcular los mejores coeficientes de un modelo de regresión lineal

#np.linalg.inv(X_1.T.dot(X_1)) calcula la inversa de la matriz resultante de la multiplicación de la transpuesta de X_1 con X_1

#.dot(X_1.T) multiplica el resultado por la transpuesta de X_1

#.dot(y) multiplica el resultado por el vector y, que son los valores objetivos de la regresión


array([1.78947368, 6.4962406 ])

# FORMA 3 - SKLEARN

Todos sabeis que si entrenamos un modelo de linear regression con sklearn podemos obtener los coeficientes con el metodo `.coef_` y el intercepto con `.intercept_`.

Haz un fit del modelo con tus datos X e y, y comprueba que valores de coeficiente (m) y que valor de intercepto (c) te da



In [38]:
X = pd.DataFrame(np.array([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20])) #Forma sklearn, la mejor forma!
y = np.array([10,10,30,40,20,40,60,40,60,80,70,60,80,100,90,100,120,130,140,120])

print("Features shape:", X.shape) #Si sabes el shape lo sabes todo
print("Target shape:", y.shape)

Features shape: (20, 1)
Target shape: (20,)


In [39]:
from sklearn.linear_model import LinearRegression #Importamos el modelo de regresion linear

lr = LinearRegression() #Lo inicializamos

In [40]:
lr.fit(X, y) #Lo entrenamos

LinearRegression()

In [41]:
print(lr.intercept_) #c

1.7894736842105061


In [42]:
print(lr.coef_) #m

[6.4962406]


# CIERRE

Si lo habeis hecho bien, los tres metodos os han debido dar valores para m y c que deberian ser identicos o suuuper parecidos. Haz un print de los parametros en una tabla para compararlos y verificar que son iguales, como se muestra a continuación.

<center>

| **Variable** | **Gradient descendt** | **Ecuacion normal** | **sklearn** |
|-------------|-------------------|----------------------|---------------------|
|m    |  6.496240601503783  |  6.4962406  |  6.4962406  |
|c |  1.7894736842102228  |  1.78947368  |  1.7894736842105061  |


y=m * x + c

</center>